src/Asm89/Stack/CorsService.php

```diff
@@ -72,134 +67,159 @@ public function isPreflightRequest(Request $request)
             && $request->headers->has('Access-Control-Request-Method');
```

Comment:
Calling this method requires adding `Vary: Origin, Access-Control-Request-Method`, but doing so will add it to every request.

To avoid that, maybe refactor to something like this:

```php
// Cache is already varied by method.
if ($request->getMethod() !== 'OPTIONS) {
  return false;
}

$this->addVary($request, 'Origin, Access-Control-Request-Method');
return $this->isCorsRequest($request) && $request->headers->has('Access-Control-Request-Method');
```
which will only `Vary` the `OPTIONS` requests (which is way better, imho).

Reply:
or even better would be to change this in `Cors::handle()` like this:
```php
if ($request->getMethod() === 'OPTIONS') {
  if ($this->cors->isPreflightRequest($request)) {
   $this->cors->handlePreflightRequest($request);
  } else {
   $response = $this->app->handle($request, $type, $catch);
 }
   $this->addVary($response, 'Origin, Access-Control-Request-Method');
}
```
This way it's either taken over, or passed, but regardless, get the `Vary` header added.

Reply:
Basically we would either need to Vary origin for ALL responses, or remove the cors check everywhere?
Maybe we could just add Vary origin in the middleware always, and remove the check if the methods header is set?

Reply:
> Basically we would either need to Vary origin for ALL responses, or remove the cors check everywhere?

So what happens if a request gets sent like this:
```
OPTIONS https://example.com/
```

And then later:
```
OPTIONS https://example.com/
Origin: https://example.org/
```

And then later:
```
OPTIONS https://example.com/
Origin: https://example.org/
Access-Control-Request-Method: PATCH
```

Without adding `Vary: Origin, Access-Control-Request-Method` to all of the `OPTIONS` requests, the first would be cached, the second and third responses would result in the same response as the first and the last would be missing the Preflight headers. :(

> Maybe we could just add Vary origin in the middleware always, and remove the check if the methods header is set?

I think you may have that backwards, whatever we are checking for first needs to be in the `Vary` for all requests. If the thing we fail on is `Access-Control-Request-Headers` instead of `Origin` then we would need to add:
`Vary: Access-Control-Request-Method` to all `OPTIONS` requests (which would make the above example pass). We could use that header as a way to determine if it's a CORS request or not. This would result in fewer cache variants than varying by `Origin` (which isn't always necessary anyways). If we then proceed to the origin check, we would need to append `Vary: Origin` to all OPTIONS requests where `Vary: Access-Control-Request-Method` is set.

The reason I would put `Vary: Origin, Access-Control-Request-Method` with the current Origin check is because of the way the condition is ordered:
https://github.com/asm89/stack-cors/blob/df153ceda9c83f6fe497907f72ab4f29c19c9b44/src/Asm89/Stack/CorsService.php#L70-L72
The first thing that is checked is the `Origin` rather than `Access-Control-Request-Method`. If we were to first check for `Access-Control-Request-Method` then we could simply add: `Vary: Access-Control-Request-Method` to all `OPTIONS` requests (and only add the `Vary: Origin` depending on the config like the main requests).

Though I don't think varying by the `Origin` is always necessary (as we've seen from the main request).

Like the primary requests, there isn't technically a need to check that it's a CORS request at all, as I've shown here:
https://github.com/davidbarratt/stack-cors/blob/22e70f5d51903f731e4608b7061e208d56f83db4/src/Asm89/Stack/Cors.php#L48-L61

That removes the CORS check completely, but the downside is, is that the application may not handle the request, if it doesn't we should provide a (static, if possible) response anyways.

The upside is that it increases cachability of the preflight because (depending on the config) you could get a completely static response that could be cached for a long time.

The only downside is if the `OPTIONS` request throws an error (rather than an simply not implemented) it would be hard to see the error. Though I suppose we could append a custom header with the error message or something so it's not completely opaque.